src/flb_ra_key.c

```diff
@@ -150,40 +154,41 @@ static int subkey_to_object(msgpack_object *map, struct mk_list *subkeys,
             }
 
             /* Index limit and ensure no overflow */
-            if (entry->array_id == INT_MAX ||
-                cur.via.array.size < entry->array_id + 1) {
+            if (entry->array_id >= cur.via.array.size) {
                 return -1;
             }
 
             val = &cur.via.array.ptr[entry->array_id];
             cur = *val;
             key = NULL; /* fill NULL since the type is array. */
-            goto next;
+            matched++;
+
+            if (levels == matched) {
+                break;
+            }
+
+            continue;
         }
 
+        /* Handle map objects */
         if (cur.type != MSGPACK_OBJECT_MAP) {
```

Comment:
in this case, what will happen if the accessed value is another array ?

Reply:
Good question. When the accessed value is another array, the function will process it in the next iteration through if (entry->type == FLB_RA_KEY_ARRAY), which runs before the map type check. This is validated in cb_nested_array_access() test, which successfully accesses $matrix[1][2] from {"matrix": [[1, 2, 3], [4, 5, 6], [7, 8, 9]]} to get the value 6.